lib/semantics/expression.cc

```diff
@@ -2390,8 +2362,84 @@ MaybeExpr ExpressionAnalyzer::MakeFunctionRef(
     return std::nullopt;
   }
 }
+
+void ArgumentAnalyzer::Analyze(
+    const parser::ActualArgSpec &arg, bool isSubroutine) {
+  // TODO: C1002: Allow a whole assumed-size array to appear if the dummy
+  // argument would accept it.  Handle by special-casing the context
+  // ActualArg -> Variable -> Designator.
+  // TODO: Actual arguments that are procedures and procedure pointers need to
+  // be detected and represented (they're not expressions).
+  // TODO: C1534: Don't allow a "restricted" specific intrinsic to be passed.
+  std::optional<ActualArgument> actual;
+  std::visit(
+      common::visitors{
+          [&](const common::Indirection<parser::Expr> &x) {
+            // TODO: Distinguish & handle procedure name and
+            // proc-component-ref
+            actual = Analyze(x.value());
+          },
+          [&](const parser::AltReturnSpec &) {
+            if (!isSubroutine) {
+              context_.Say("alternate return specification may not appear on"
+                           " function reference"_err_en_US);
+            }
+          },
+          [&](const parser::ActualArg::PercentRef &) {
+            context_.Say("TODO: %REF() argument"_err_en_US);
+          },
+          [&](const parser::ActualArg::PercentVal &) {
+            context_.Say("TODO: %VAL() argument"_err_en_US);
+          },
+      },
+      std::get<parser::ActualArg>(arg.t).u);
+  if (actual.has_value()) {
+    if (const auto &argKW{std::get<std::optional<parser::Keyword>>(arg.t)}) {
+      actual->keyword = argKW->v.source;
+    }
+    actuals_.emplace_back(std::move(*actual));
+  } else {
+    success_ = false;
+  }
+}
+
+std::optional<ActualArgument> ArgumentAnalyzer::Analyze(
+    const parser::Expr &expr) {
+  if (const Symbol * assumedTypeDummy{AssumedTypeDummy(expr)}) {
+    return ActualArgument{ActualArgument::AssumedType{*assumedTypeDummy}};
+  } else if (MaybeExpr argExpr{context_.Analyze(expr)}) {
+    Expr<SomeType> x{Fold(context_.GetFoldingContext(), std::move(*argExpr))};
+    if (const auto *proc{std::get_if<ProcedureDesignator>(&x.u)}) {
+      if (!std::holds_alternative<SpecificIntrinsic>(proc->u) &&
+          proc->IsElemental()) {  // C1533
+        context_.Say(expr.source,
+            "Non-intrinsic ELEMENTAL procedure cannot be passed as argument"_err_en_US);
+      }
+    }
+    if (auto coarrayRef{ExtractCoarrayRef(x)}) {
+      const Symbol &coarray{coarrayRef->GetLastSymbol()};
+      if (const semantics::DeclTypeSpec * type{coarray.GetType()}) {
+        if (const semantics::DerivedTypeSpec * derived{type->AsDerived()}) {
+          if (auto ptr{semantics::FindPointerUltimateComponent(*derived)}) {
+            if (auto *msg{context_.Say(expr.source,
```

Comment:
If we're emitting an error message, should we be returning ```std::nullopt```?

Reply:
I'm not sure. This code isn't new, it's just been moved.